projects/plugins/boost/app/modules/cache/Page_Cache_Setup.php

```diff
@@ -0,0 +1,186 @@
+<?php
+
+namespace Automattic\Jetpack_Boost\Modules\Page_Cache;
+
+class Page_Cache_Setup {
+
+	/**
+	 * Runs setup steps and returns whether setup was successful or not.
+	 * @return bool
+	 */
+	public static function run_setup() {
+		$steps = array(
+			'is_wp_content_writable',
+			'create_advanced_cache',
+			'add_wp_cache_define',
+		);
+
+		foreach ( $steps as $step ) {
+			$result = self::$step();
+			if ( is_wp_error( $result ) ) {
+				jetpack_boost_ds_set( 'page_cache_error', $result->get_error_code() );
+
+				return false;
+			}
+		}
+
+		jetpack_boost_ds_set( 'page_cache_error', '' );
+
+		return true;
+	}
+
+	/**
+	 * Returns true if the wp-content directory is writeable.
+	 */
+	private static function is_wp_content_writable() {
+		$filename = WP_CONTENT_DIR . '/' . uniqid() . '.txt';
+		$result   = file_put_contents( $filename, 'test' ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_file_put_contents
+		wp_delete_file( $filename );
+
+		if ( $result === false ) {
+			return new \WP_Error( 'wp-content-not-writable', 'wp-content directory is not writeable' );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Creates the advanced-cache.php file.
+	 *
+	 * Returns true if the files were setup correctly, or WP_Error if there was a problem.
+	 *
+	 * @return bool|WP_Error
+	 */
+	private static function create_advanced_cache() {
+		$advanced_cache_filename = WP_CONTENT_DIR . '/advanced-cache.php';
+
+		if ( file_exists( $advanced_cache_filename ) ) {
+			$content = file_get_contents( $advanced_cache_filename ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
+
+			if ( strpos( $content, Page_Cache::ADVANCED_CACHE_SIGNATURE ) === false ) {
+				return new \WP_Error( 'advanced-cache-incompatible', 'advanced-cache.php exists, but belongs to another plugin/system.' );
```

Comment:
I like that these WP_Errors each have a code that let our UI describe what's going on in detail.